### ISSUE TITLE
Improvements to standalone forms

### ISSUE THREAD

**Issue by danblah** [HyphaApp/hypha#947]
There are a number of issues with the current standalone forms implementation. I'm strongly recommending pulling it from live until we've got these. This is a priority issue that is required for us to cease using the old submission application's standalone forms.

**Required functionality for us to leave the old app:**
- [x] They have no styling. Have them take on the same styling as the fund/lab forms (considering this a bug)
- [ ] Add the ability to restrict form access to specified roles in the form editor. We need forms that are only available to the team, applicants, reviewers (and later for approved applicants from specific funds).
- [x] Require a recaptcha (or some other anti-spam) field for forms that are publically available. 
- [ ] File attachment field
- [x] A basic interface to review submissions (maybe this exists, can't tell cause currently erroring on test submissions)
- [ ] Field to define team lead
- [ ] Field to define slack channel to send notifications
- [ ] Notification to site-wide slack channel on all submissions
- [ ] Field to define webhook endpoint to send submissions to
- [x] Hide none wagtail forms from /admin/forms/
- [ ] Ability to preview completed form before submitting

**Comment by frjo** [comment]
Related PR is https://github.com/OpenTechFund/opentech.fund/pull/948

**Comment by frjo** [comment]
Since this is behind Cloudflare I do not think a captcha is needed. And if there are spam submissions we can do better than annoying captchas.

What is the "Team lead" field for. Just to direct notification?

The "webhook" field should end up in a separate issue I think, as an enhancement. I assume the system should not only ping the url put send some payload as well?

**Comment by frjo** [comment]
@Parbhat Can you take on the following tasks?

1. Add a file upload field to the available fields when you build a form.
2. Make it possible to restrict access to a form by role/group.

**Comment by frjo** [comment]
@danblah Wagtails forms on live is being blocked by Cloudflare cache, we have a page rule that tells Cloudflare to cache everything on www.opentech.fund. (By default Cloudflare only cache static assets like css, js and images, not HTML. So the cache bypass rules for test/dev/live is not really needed I think).

To get a page rule to bypass cache for pages with a cookie you would need to upgrade to a Cloudflare biz account. Overkill, and I think Cloudflare gets enough money already.

Could we decide that all wagtail forms get "-form" as part of their slug? I have added a page rule now that bypass the cache for "https://www.opentech.fund/*-form*".

The https://www.opentech.fund/about/jobs/test-form/ form seems to work now. I have also deployed the template fixes so the form looks like other forms on the site.

**Comment by Parbhat** [comment]
@frjo Thanks for fixing templates. Will work on the other requirements. Need some feedback on the second point.

>Make it possible to restrict access to a form by role/group.

Do we need a form like [this](https://www.opentech.fund/about/jobs/test-form/) test form to be visible only to a certain type of users and not public? What should be the desired workflow? We can make a page accessible to users in specific groups using Private pages feature https://docs.wagtail.io/en/v2.4/advanced_topics/privacy.html

**Comment by frjo** [comment]
@Parbhat Private pages sounds like it should work. Nice that it's built it to wagtail.

OTF want to set up forms that e.g  only staff should have access to and others that only reviewers should see. It can be request for expense reimbursement for staff, questioners for reviewers etc.

Please notice that I added a third feature to your list above, about hiding none wagtail forms from the /admin/forms/ page.

**Comment by danblah** [comment]
> What should be the desired workflow? 
@Parbhat, ideally the form author can select which group/role they want to give access to in the form creation/edit page (for example https://apply.opentech.fund/admin/pages/814/edit/).

**Comment by danblah** [comment]
> What is the "Team lead" field for. Just to direct notification?

@frjo, that's right. this would enable a notification to the site-wide channel in the same format as others, for example:
@lead A new submission has been submitted for <name/link to form>

> The "webhook" field should end up in a separate issue I think, as an enhancement. I assume the system should not only ping the url put send some payload as well?

@frjo, that's right. the object/payload should have a name/value for as many of the fields being set in the edit page as possible.

**Comment by danblah** [comment]
> 1. Add a file upload field to the available fields when you build a form.

@Parbhat and @frjo, want to be sure these are private files, only accessible to defined roles (or team only if that's easier right now).

**Comment by danblah** [comment]
> 3. Hide none wagtail forms from /admin/forms/

@Parbhat, plus one to that page only showing wagtail forms (hiding all the ones created from the apply app).

**Comment by Parbhat** [comment]
@frjo For the file field upload option, should we allow to upload only images or other types of files as well? Wagtail forms do not have file upload feature but we can create it. Apply site has a [custom](https://github.com/OpenTechFund/opentech.fund/tree/master/opentech/apply/stream_forms) implementation of forms using stream field. Do we want to look into the options to use stream field forms here as well instead of wagtail forms?

Created PR #963 to hide apply forms from `/admin/forms/`

**Comment by danblah** [comment]
> @frjo For the file field upload option, should we allow to upload only images or other types of files as well? Wagtail forms do not have file upload feature but we can create it. Apply site has a [custom](https://github.com/OpenTechFund/opentech.fund/tree/master/opentech/apply/stream_forms) implementation of forms using stream field. Do we want to look into the options to use stream field forms here as well instead of wagtail forms?
> 
> Created PR #963 to hide apply forms from `/admin/forms/`

**Comment by danblah** [comment]
Whoops, accidentally closed. 

> > @frjo For the file field upload option, should we allow to upload only images or other types of files as well?

.md, .rtf, .txt, and all office and libreoffice formats + pdf, jpg and png.

**Comment by Parbhat** [comment]
Thanks @danblah

**Comment by Parbhat** [comment]
PR for 1. Add a file upload field to the available fields when you build a form. - #976 

We will probably need an interface to review form submissions. As the default form submissions interface (In Wagtail admin sidebar), only displays text data. Currently storing the file name of uploaded file with every submission

**Comment by fourthletter** [comment]
@frjo Could we also remove the requirement that forms should include certain fields? Users prefer more flexibility in how they design their forms/fund applications.

**Comment by frjo** [comment]
@fourthletter This is not the application forms, it's the standalone forms. Not mush used I believe.

**Comment by Erioldoesdesign** [comment]
> This is not the application forms, it's the standalone forms. Not mush used I believe.

@frjo Am i correct in assuming these stand-alone forms are no longer used/rarely used therefore the improvements listed in checkboxes above are very not urgent?

I'm also assuming that 'stand alone forms' are application forms that are **not** connected to a fund/lab and are kind of like 'flash grants' @danblah have I made the correct(ish)assumption here? if so I'll update the issue body with some context.

**Comment by Erioldoesdesign** [comment]
Q1 2022 Rebasing updates:
* Cleared old Milestone
* Removed priority needed label
* Asked fredrik how critical this is since it's 'not much used' and appears to have not been worked on since 2020.
* Asked Dan for clarified context -> Needs a better title and issue description
* Recommend review in Q2 rebasing for priority level.
* Size estimated at M
* Removed from dev backlog board: https://github.com/HyphaApp/hypha/projects/2

**Comment by frjo** [comment]
This is part of the public side of Hypha. A part soon no one will be using. I'm closing it and recommend we close all issues realting to the public side.
